### PR TITLE
[tf.data] Check in a strictly faster rejection resampling transformation.

### PR DESCRIPTION
**Pull request by joel-shor**
This transformation is faster because it rejects fewer data. This
is done by occasionally sampling from the original data distribution
in an efficient way.

Tested:
bazel test :resample_test

### CODE REVIEW

File: tensorflow/contrib/data/python/kernel_tests/resample_test.py [hunk 1]
```diff
@@ -109,5 +134,55 @@ def _remap_fn(_):
...
+
+    fast_time = _time_resampling(self, data_np, target_dist, init_dist,
+                                 use_v2=True, num_to_sample=num_to_sample)
+    slow_time = _time_resampling(self, data_np, target_dist, init_dist,
+                                 use_v2=False, num_to_sample=num_to_sample)
+
+    self.assertLess(fast_time, slow_time)
+
+
+class MapDatasetBenchmark(test.Benchmark):
```
Comment (mrry) [inline tensorflow/contrib/data/python/kernel_tests/resample_test.py]: Nit: s/MapDataset/Resample/

File: tensorflow/contrib/data/python/ops/resampling.py [hunk 2]
```diff
@@ -50,70 +51,171 @@ def rejection_resample(class_func, target_dist, initial_dist=None, seed=None):
...
       acceptance_dist_ds = initial_dist_ds.map(
           lambda initial: _calculate_acceptance_probs(initial, target_dist_t))
+    return _filter_ds(dataset, acceptance_dist_ds, initial_dist_ds,
+                      class_values_ds, seed)
+
+  return _apply_fn
+
+
+def rejection_resample_v2(class_func, target_dist, initial_dist=None,
+                          seed=None):
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/resampling.py]: Should we just replace the old version with this, and call it `rejection_resample()`?

File: tensorflow/contrib/data/python/ops/resampling.py [hunk 3]
```diff
@@ -50,70 +51,171 @@ def rejection_resample(class_func, target_dist, initial_dist=None, seed=None):
...
+  """A transformation that resamples a dataset to achieve a target distribution.
+
+  This differs from v1 in that it will also sample from the original dataset
+  with some probability, so it makes strictly fewer data rejections. Due to an
+  implementation detail it must initialize a separate dataset initializer, so
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/resampling.py]: I think this is no longer true, is it?

**Review by mrry — DISMISSED** [review]
New version looks good to me. Just a few minor comments and we should be good to merge.

**Comment by ekelsen** [comment]
@joel-shor Could you address the comments and resolve the conflicts?

**Comment by joel-shor** [comment]
Applied changes.

Note that I also added one bit of extra logic, which short-circuits the interleaving (making it faster) and avoiding a bug in one of those edge cases.

**Review by joel-shor — COMMENTED** [review]
Done.

File: tensorflow/contrib/data/python/ops/resampling.py [hunk 4]
```diff
@@ -50,79 +51,184 @@ def rejection_resample(class_func, target_dist, initial_dist=None, seed=None):
...
 
   return _apply_fn
 
 
-def _calculate_acceptance_probs(initial_probs, target_probs):
-  """Calculate the per-class acceptance rates.
+def _get_prob_original_static(initial_dist_t, target_dist_t):
+  """Returns the static probability of sampling from the original.
+
+  For some reason, `tensor_util.constant_value(prob_of_original)` of a ratio
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/resampling.py]: > For some reason, [...].

The reason is that `tensor_util.constant_value()` only implements a handful of ops, and returns `None` when it encounters an op that it hasn't implemented. Please update the comment to make it seem less mysterious!

File: tensorflow/contrib/data/python/kernel_tests/resample_test.py [hunk 5]
```diff
@@ -30,52 +33,97 @@
...
+      sess.run(get_next)
+    end_time = time.time()
+
+  return end_time - start_time
+
+
+class ResampleTest(test.TestCase, parameterized.TestCase):
+
+  @parameterized.named_parameters(
+      ("InitialnDistributionKnown", True),
```
Comment (mrry) [inline tensorflow/contrib/data/python/kernel_tests/resample_test.py]: Typo: "Initialn" -> "Initial"

**Review by mrry — CHANGES_REQUESTED** [review]
Just a couple of minor changes left.

File: tensorflow/contrib/data/python/ops/resampling.py (hunk 4 above)
Comment (joel-shor) [reply tensorflow/contrib/data/python/ops/resampling.py]: Done

File: tensorflow/contrib/data/python/kernel_tests/resample_test.py (hunk 5 above)
Comment (joel-shor) [reply tensorflow/contrib/data/python/kernel_tests/resample_test.py]: Done

**Review by joel-shor — COMMENTED** [review]
Done

**Comment by joel-shor** [comment]
Done! PTAL

**Comment by mrry** [comment]
Looks like a legitimate failure on the Windows build. The usual cause for errors like that is the fact that `np.array([1]).dtype` (for any list of Python integers) is `int64` on Linux and `int32` on Windows. That probably means that some integer array created in the code needs to be explicitly cast to `int64`.

I’d start by looking at arguments to `sample_from_datasets()` although it’s possible that the bug is in the Python implementation of that function.

File: tensorflow/contrib/data/python/ops/interleave_ops.py [hunk 6]
```diff
@@ -240,9 +240,9 @@ def select_dataset(logits, seed):
...
 
-  logging.warn('selector_input.output_types: %s', selector_input.output_types)
-  logging.warn('selector_input.output_shapes: %s', selector_input.output_shapes)
+  logging.warn('selector_input.output_types: %s', str(selector_input.output_types))
+  logging.warn('selector_input.output_shapes: %s', str(selector_input.output_shapes))
   for i, dataset in enumerate(datasets):
-    logging.warn('dataset %i output_types: %s' % (i, dataset.output_types))
-    logging.warn('dataset %i output_shapes: %s' % (i, dataset.output_shapes))
+    logging.warn('dataset %i output_types: %s' % (i, str(dataset.output_types)))
+    logging.warn('dataset %i output_shapes: %s' % (i, str(dataset.output_shapes)))
```
Comment (mrry) [inline tensorflow/contrib/data/python/ops/interleave_ops.py]: (Not sure if you're still debugging this, but...) Can you please remove these `logging.warn()` calls before submitting?

Comment (joel-shor) [reply tensorflow/contrib/data/python/ops/interleave_ops.py]: I am still debugging...I've just verified that the types and shapes of the datasets to `DirectedInterleaveDataset` are now the same between Windows and Linux, but I'm still getting the `kernel not registered` error. Any other suggestions for what might be causing this?

File: tensorflow/contrib/data/python/kernel_tests/resample_test.py [hunk 7]
```diff
@@ -109,5 +158,23 @@ def _remap_fn(_):
...
+
+class ResampleDatasetBenchmark(test.Benchmark):
+
+  def benchmarkResamplePerformance(self):
+    init_dist = [0.25, 0.25, 0.25, 0.25]
+    target_dist = [0.0, 0.0, 0.0, 1.0]
+    num_classes = len(init_dist)
+    # We don't need many samples to test a dirac-delta target distribution
+    num_samples = 1000
+    data_np = np.random.choice(num_classes, num_samples, p=init_dist)
```
Comment (mrry) [inline tensorflow/contrib/data/python/kernel_tests/resample_test.py]: This might be `np.int32` on Windows.

**Comment by joel-shor** [comment]
@mrry I think the problem was that the interleave dataset op wasn't properly added to the cmake file. PTAL when you get a chance, once the tests are completed.

**Comment by mrry** [comment]
Ah, good catch! I guess the `interleave_dataset_op_test.py` was [disabled due to an unrelated bug](https://github.com/tensorflow/tensorflow/blob/d0f5bc17560fc97bcc7de9164aa3b237a8d5221d/tensorflow/contrib/cmake/tf_tests.cmake#L285) and this left a hole in the test coverage.

**Review by mrry — APPROVED** [review]

**Pull request merged**